docs/pallas/tpu/sparse.md

```diff
@@ -0,0 +1,571 @@
+  timestamp: 1726001133029
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: ibeIs_6QFMAM
+outputId: d72edb91-4529-4650-c9e9-b96788608635
+---
+import functools
+import timeit
+import numpy as np
+import jax
+from jax import numpy as jnp
+from jax import lax
+from jax.experimental import checkify
+from jax.experimental import pallas as pl
+from jax.experimental.pallas import tpu as pltpu
+
+assert "TPU" in jax.devices()[0].device_kind, "Please run this notebook with TPU devices."
+print("Running on", jax.devices()[0].device_kind)
+```
+
++++ {"id": "FIDGpPTEIcOa"}
+
+## Dynamic Block Indexing with Scalar Prefetch
+
+We will be exploiting the "scalar prefetch" feature of Pallas to enable us to write sparse kernels. Scalar prefetch allows you to pass in a small amount of data into SMEM ("scalar") that is loaded before the invocation of the kernel ("prefetch"). This metadata is then passed into the `index_map` for each BlockSpec, allowing the you to perform data-dependent indexing calculations. The main goal of this tutorial is to go over common programming patterns that utilize this feature.
+
+To enable scalar prefetch, use the `PrefetchScalarGridSpec` in place of the standard `pl.GridSpec`:
+
+```python
+class PrefetchScalarGridSpec:
+  def __init__(self,
+    num_scalar_prefetch: int,
+    grid: tuple[int, ...],
+    in_specs: PyTree[BlockSpec],
+    out_specs: PyTree[BlockSpec],
+    scratch_shapes: tuple[MemorySpace, ...]):
+      ...
+```
+
+There are two additional agruments for `pltpu.PrefetchScalarGridSpec` compared to `pl.GridSpec`:
+- `num_scalar_prefetch`, which indicates the number of scalar prefetch values.
+- `scratch_shapes`, which allocates a list of temporary "scratchpad" Refs for storing intermediate calculations (generally useful but unrelated to this tutorial).
+
+We are primarily interested in `num_scalar_prefetch`. When this is set to a non-zero value, it changes the call signature of the kernel and index maps to expect additional prefetch values. The prefetch `Ref`s passed in to the `index_map` and kernel are all allocated in SMEM and are not partitioned into blocks as they do not have a BlockSpec defined. Moreover, the order of arguments to both `index_map` and kernel are always fixed and described below:
+
+- Each `BlockSpec`'s `index_map` now expects the prefetch `Ref`s to come after the grid indices:
+```python
+def index_map(*grid_indices, *prefetch_refs):
+    ...
+```
+
+- The user-defined kernel expects prefetch `Ref`s to come before the input `Ref`s. Additionally, the scratch refs come after the output `Ref`s.
+```python
+def kernel(*prefetch_refs, *input_refs, *output_refs, *scratch_refs):
+    ...
+```
+
+- When calling a new kernel using `pallas_call`, the function returned by `pallas_call` also expects the scalar prefetch arguments to come before the inputs, e.g.
+```python
+kernel = pl.pallas_call(...)
+result = kernel(*prefetch_args, *input_args)
+```
+
```

Comment:
Can you add a note about Pytrees here? 

I.e. scalar prefetch values can be Pytrees. All of of their leaf values will be prefetched together and `num_scalar_prefetch` really refers to the number of arguments (i.e. pytrees) not the number of array values.